### PR TITLE
Database api with entities, table creation sql query, hibernate cfg and etc

### PR DESCRIPTION
**Pull request by YuriiFilatov**

### CODE REVIEW

File: server/src/main/java/ru/hse/database/dao/DaoChapter.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DaoChapter {
+
+    static Logger logger = LoggerFactory.getLogger(DaoChapter.class);
+
+    static public Chapter getChapterById(long id) {
+        Session session = HibernateUtil.getSessionFactory().openSession();
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/dao/DaoChapter.java]: better to use try with resources

File: server/src/main/java/ru/hse/database/dao/DaoChapter.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        Group group;
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            chapter = session.get(Chapter.class, chapter.getChapterId());
+
+            session.beginTransaction();
+            group = chapter.getGroupHost();
+            session.getTransaction().commit();
+        }
+
+        return group;
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/dao/DaoChapter.java]: why not create group on initialize and do all in try block?

File: server/src/main/java/ru/hse/database/dao/DaoChapter.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    static public void createOrUpdateChapter(Chapter chapter) {
+        try {
+            try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+                session.beginTransaction();
+                session.merge(chapter);
+                session.getTransaction().commit();
+            }
+        } catch (Exception e) {
+            logger.error("Error while merging chapter with id " + chapter.getChapterId(), e);
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/dao/DaoChapter.java]: mb use {} in format string like in printf

File: server/src/main/java/ru/hse/database/dao/DaoGroup.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    static Logger logger = LoggerFactory.getLogger(DaoGroup.class);
+    static public Group getGroupById(long id) {
+        Session session = HibernateUtil.getSessionFactory().openSession();
+
+        Group group = session.get(Group.class, id);
+
+        session.close();
+        return group;
+    }
+    static public List<Group> getGroupsByName(String goupName) {
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/dao/DaoGroup.java]: misprint groupName

File: server/src/main/java/ru/hse/database/dao/DaoChapter.java (hunk 2 above)
Comment (Redvin-dt) [reply server/src/main/java/ru/hse/database/dao/DaoChapter.java]: and same in other places

File: server/src/main/java/ru/hse/database/dao/DaoUser.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            users = query.getResultList();
+        }
+        if (users.isEmpty()) {
+            return null;
+        }
+
+        if (users.size() > 1) {
+            logger.error("More than 1 users in table with user login " + login);
+            return null;
+        }
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/dao/DaoUser.java]: mb throw an error to catch it in future and fix table state?

File: server/src/main/java/ru/hse/database/dao/DaoChapter.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    static public void deleteChapter(Chapter chapter) {
+        Long id = chapter.getChapterId();
+        try (Session session = HibernateUtil.getSessionFactory().openSession()){
+            session.beginTransaction();
+            session.remove(chapter);
+            session.getTransaction().commit();
+        } catch (Exception e) {
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/dao/DaoChapter.java]: catch more specific exception, IllegalStateException

File: server/src/main/java/ru/hse/database/entities/Chapter.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    @Setter
+    @Getter
+    @Column(name = "meta_info")
+    private String metaInfo;
+
+    @Setter
+    @Getter
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "group_host")
+    private Group groupHost;
```
Comment (Redvin-dt) [inline server/src/main/java/ru/hse/database/entities/Chapter.java]: mb set Getter and Setter before class declaration and set Setter(AccessLevel.NONE),  Getter(AccessLevel.NONE) to id

Comment (Redvin-dt) [reply server/src/main/java/ru/hse/database/entities/Chapter.java]: and same in other classes

**Review by Redvin-dt — APPROVED** [review]

File: server/src/main/java/ru/hse/database/utils/HibernateUtil.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            sessionFactory = new MetadataSources(registry).buildMetadata().buildSessionFactory();
+        } catch (Exception e) {
+            StandardServiceRegistryBuilder.destroy(registry);
+            throw new ExceptionInInitializerError("Initial SessionFactory failed" + e);
+        }
+        return sessionFactory;
+    }
+
+    public static void shutdown() {
+        getSessionFactory().close();
```
Comment (devcrocod) [inline server/src/main/java/ru/hse/database/utils/HibernateUtil.java:29]: can then make this class as closeable or autocloseable?

File: server/src/main/java/ru/hse/database/utils/HibernateUtil.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package ru.hse.database.utils;
+
+import lombok.Getter;
+import org.hibernate.SessionFactory;
+import org.hibernate.boot.MetadataSources;
+import org.hibernate.boot.registry.StandardServiceRegistry;
+import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
+
+public class HibernateUtil {
```
Comment (devcrocod) [inline server/src/main/java/ru/hse/database/utils/HibernateUtil.java]: Is this an abstract factory?

File: server/src/main/java/ru/hse/database/dao/DaoUser.java [hunk 10]
```diff
@@ -30,7 +30,7 @@ static public User getUserById(long id) {
             return session.get(User.class, id);
         }
     }
-    public User getUserByLogin(String login) throws NotUniqueUserLoginException {
+    static public User getUserByLogin(String login) throws NotUniqueUserLoginException {
```
Comment (devcrocod) [inline server/src/main/java/ru/hse/database/dao/DaoUser.java:33]: Why static?

**Pull request merged**